c/src/nnstreamer-capi-pipeline.c

```diff
@@ -1480,11 +1483,17 @@ ml_pipeline_src_input_data (ml_pipeline_src_h h, ml_tensors_data_h data,
     /** @todo Verify that gst_buffer_append lists tensors/gstmem in the correct order */
   }
 
+  /* Unlock if it's not auto-free. We do not know when it'll be freed. */
```

Comment:
line 1486 ~ 1515
1. deadlock case if data is invalid 
2. If given policy is do-not-free case, buffers are wrapped and pushed into pipeline. So I think this unlock code (1486 ~1488) is unnecessary.

how about adding unlock in line 1511?
```
destroy_data:
  if (_data != NULL) {
    G_UNLOCK_UNLESS_NOLOCK (*_data);

    /* Free data handle */
    if (policy == ML_PIPELINE_BUF_POLICY_AUTO_FREE)
      ml_tensors_data_destroy (data);
  }

  handle_exit (h);
}
```

Reply:
> 1. deadlock case if data is invalid

Thanks for finding out the deadlock!

> 2. If given policy is do-not-free case, buffers are wrapped and pushed into pipeline. So I think this unlock code (1486 ~1488) is unnecessary.

If we unlock at destroy_data, the user's thread may call ```free(_data)``` before G_UNLOCK_UNLESS_NOLOCK. Thus, we need to UNLOCK before calling push_buffer.